subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/ConsoleJvmTestLoggingFunctionalTest.groovy

```diff
@@ -0,0 +1,205 @@
+            """
+        }
+
+        when:
+        succeeds(TEST_TASK_NAME)
+
+        then:
+        def groupedTaskOutput = parseTaskOutput(output, TEST_TASK_GROUP_HEADER, BUILD_SUCCESSFUL_OUTPUT)
+        groupedTaskOutput.contains("""MyTest > testExpectation ${TestLogEvent.STANDARD_OUT.consoleMarker}
+    standard output""")
+        groupedTaskOutput.contains("""MyTest > testExpectation ${TestLogEvent.STANDARD_ERROR.consoleMarker}
+    standard error""")
+    }
+
+    def "can group multiple test log events with task"() {
+        executer.withStackTraceChecksDisabled()
+
+        given:
+        buildFile << testLoggingEvents(TestLogEvent.STARTED.testLoggingIdentifier, TestLogEvent.FAILED.testLoggingIdentifier)
+        buildFile << testLoggingStandardStream()
+
+        file(JAVA_TEST_FILE_PATH) << javaTestClass {
+            """
+                System.out.println("standard output");
+                throw new RuntimeException("expected");
+            """
+        }
+
+        when:
+        fails(TEST_TASK_NAME)
+
+        then:
+        def groupedTaskOutput = parseTaskOutput(output, TEST_TASK_GROUP_HEADER, '1 test completed, 1 failed')
+        groupedTaskOutput.contains("""MyTest > testExpectation ${TestLogEvent.STANDARD_OUT.name()}
+    standard output""")
+        assertTaskOutput(groupedTaskOutput, testLogEventRegex(TestLogEvent.STARTED.consoleMarker))
+        assertTaskOutput(groupedTaskOutput, testLogEventRegex(TestLogEvent.FAILED.consoleMarker))
+    }
+
+    static String javaProject() {
+        """
+            apply plugin: 'java'
+            
+            repositories {
+                mavenCentral()
+            }
+            
+            dependencies {
+                testCompile 'junit:junit:4.12'
+            }
+        """
+    }
+
+    static String testLoggingEvents(String... events) {
+        """
+            test {
+                testLogging {
+                    events ${events.collect { "'$it'" }.join(', ')}
+                }
+            }
+        """
+    }
+
+    static String testLoggingStandardStream() {
+        """
+            test {
+                testLogging {
+                    showStandardStreams = true
+                }
+            }
+        """
+    }
+
+    static String javaTestClass(Closure<String> testMethodBody) {
+        """
+            import org.junit.Test;
+
+            public class MyTest {
+                @Test
+                public void testExpectation() {
+                    ${testMethodBody()}
+                }
+            }
+        """
+    }
+
+    static void parseAndAssertTaskOutput(String output, String taskOutputStart, String taskOutputEnd, String regexToFind) {
+        def groupedTaskOutput = parseTaskOutput(output, taskOutputStart, taskOutputEnd)
+        assertTaskOutput(groupedTaskOutput, regexToFind)
+    }
+
+    static String parseTaskOutput(String output, String taskOutputStart, String taskOutputEnd) {
```

Comment:
let's get this logic on to a test fixture rather than reimplementing this logic in every test. We added logic with exactly the same intention but completely different implementation in `ConsoleTaskGroupingFunctionalTest`